File: apps/agent/src/routes/briefings.ts
import type { Request, Response, Router } from "express";
import { Router as createRouter } from "express";
import { z } from "zod";
import {
  getBriefing,
  listBriefings,
  publicBriefing,
} from "../lib/briefings.js";
import { runBriefing } from "../agent/briefing.js";

export const briefingsRouter: Router = createRouter();

briefingsRouter.get("/briefings", async (_req: Request, res: Response) => {
  try {
    const records = await listBriefings(30);
    return res.json({
      count: records.length,
      briefings: records.map(publicBriefing),
    });
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return res.status(500).json({ error: "list_failed", detail: msg });
  }
});

briefingsRouter.get("/briefings/:id", async (req: Request, res: Response) => {
  const raw = req.params["id"];
  const id = typeof raw === "string" ? raw : Array.isArray(raw) ? raw[0] : undefined;
  if (!id) return res.status(400).json({ error: "missing_id" });
  try {
    const b = await getBriefing(id);
    if (!b) return res.status(404).json({ error: "not_found" });
    return res.json(publicBriefing(b));
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return res.status(500).json({ error: "get_failed", detail: msg });
  }
});

const generateSchema = z.object({
  event_id: z.string().min(1).optional(),
  event_title: z.string().min(1).optional(),
});

briefingsRouter.post("/briefings/generate", async (req: Request, res: Response) => {
  const parsed = generateSchema.safeParse(req.body);
  if (!parsed.success) {
    return res.status(400).json({ error: "invalid_body", detail: parsed.error.flatten() });
  }
  if (!parsed.data.event_id && !parsed.data.event_title) {
    return res.status(400).json({ error: "missing_event" });
  }

  res.status(200);
  res.setHeader("Content-Type", "text/event-stream");
  res.setHeader("Cache-Control", "no-cache, no-transform");
  res.setHeader("Connection", "keep-alive");
  res.setHeader("X-Accel-Buffering", "no");
  res.flushHeaders?.();
  res.write(": connected\n\n");

  const heartbeat = setInterval(() => {
    res.write(": ping\n\n");
  }, 15000);

  let closed = false;
  res.on("close", () => {
    closed = true;
    clearInterval(heartbeat);
  });

  const write = (event: string, payload: unknown) => {
    if (closed) return;
    res.write(`event: ${event}\n`);
    res.write(`data: ${JSON.stringify(payload)}\n\n`);
  };

  try {
    const opts: { eventId?: string; eventTitle?: string } = {};
    if (parsed.data.event_id) opts.eventId = parsed.data.event_id;
    if (parsed.data.event_title) opts.eventTitle = parsed.data.event_title;
    for await (const ev of runBriefing(opts)) {
      if (closed) break;
      write(ev.kind, ev);
      if (ev.kind === "done" || ev.kind === "error") break;
    }
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    write("error", { kind: "error", message: msg, at: Date.now() });
  } finally {
    clearInterval(heartbeat);
    if (!closed) res.end();
  }
});

File: apps/agent-py/app/routes/briefings.py
"""/briefings — port of apps/agent/src/routes/briefings.ts."""
from __future__ import annotations

import json

from fastapi import APIRouter
from fastapi.responses import JSONResponse, StreamingResponse
from pydantic import BaseModel

from app.agent.briefing import run_briefing
from app.lib.briefings import get_briefing, list_briefings, public_briefing

router = APIRouter()
_SSE_HEADERS = {"Cache-Control": "no-cache, no-transform", "Connection": "keep-alive", "X-Accel-Buffering": "no"}


@router.get("/briefings")
async def list_route() -> JSONResponse:
    records = await list_briefings(30)
    return JSONResponse({"count": len(records), "briefings": [public_briefing(b) for b in records]})


@router.get("/briefings/{bid}")
async def get_route(bid: str) -> JSONResponse:
    b = await get_briefing(bid)
    if not b:
        return JSONResponse(status_code=404, content={"error": "not_found"})
    return JSONResponse(public_briefing(b))


class GenerateBody(BaseModel):
    event_id: str | None = None
    event_title: str | None = None


@router.post("/briefings/generate")
async def generate_route(body: GenerateBody) -> StreamingResponse:
    async def gen():
        yield ": connected\n\n"
        try:
            async for ev in run_briefing(event_id=body.event_id, event_title=body.event_title):
                yield f"event: {ev['kind']}\ndata: {json.dumps(ev, default=str)}\n\n"
                if ev["kind"] in ("done", "error"):
                    break
        except Exception as err:  # noqa: BLE001
            yield f"event: error\ndata: {json.dumps({'kind': 'error', 'message': str(err)})}\n\n"

    return StreamingResponse(gen(), media_type="text/event-stream", headers=_SSE_HEADERS)
